### PR TITLE
Setup End To End Tests

### PR DESCRIPTION
**Pull request by timvahlbrock**
### Description

PR to setup end to end tests.

## Checklist

- [x] My code follows the style guidelines
- [x] I have documented my code
- [x] My changes generate no new warnings
- [x] I have linked the Issue
- [x] The Pull Request covers all the Checkpoints from the tagged Issue
- [x] I have noted the hours of work i have put in the issue

## I have edited following Parts of the project

- [x] Frontend
- [x] Backend
- [x] Github
- [x] Root

## If tests are required did you make them?

- [ ] No tests required
- [x] Yes
- [ ] No

### CODE REVIEW

**Review by timvahlbrock — COMMENTED** [review]
Rechecked all files. Ready for review.

More meaningful tests are added in a further step.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -62,37 +63,39 @@ jobs:
...
 
-        - name: Build and Push
+        - name: Build and Push Container
           uses: docker/build-push-action@v4
           with:
-            context: backend
+            context: frontend
             platforms: linux/amd64,linux/arm64
-            push: ${{ github.ref == 'refs/heads/main' }}
+            push: true
```
Comment (Paulomart) [inline .github/workflows/ci.yml:90]: I think this can be omitted

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -62,37 +63,39 @@ jobs:
...
             tags: |
-              ${{ env.FRONTEND_DOCKER_HUB_IMAGE_NAME }}:latest,
-              ${{env.GITHUB_REGISTRY}}/${{ env.FRONTEND_GITHUB_IMAGE_NAME }}:latest
-            cache-from: type=gha
-            cache-to: type=gha,mode=max
-
+              ${{ env.FRONTEND_DOCKER_HUB_IMAGE_NAME }}:${{ github.sha }}-untested,
```
Comment (Paulomart) [inline .github/workflows/ci.yml:92]: I would prefer to only push to ghcr to minimize build time

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -126,45 +129,196 @@ jobs:
...
+
+      - name: Build and Push Container
+        uses: docker/build-push-action@v4
+        with:
+            context: backend
+            platforms: linux/amd64,linux/arm64
+            push: true
+            tags: |
+              ${{ env.BACKEND_DOCKER_HUB_IMAGE_NAME }}:${{ github.sha }}-untested,
+              ${{env.GITHUB_REGISTRY}}/${{ env.BACKEND_GITHUB_IMAGE_NAME }}:${{ github.sha }}-untested
```
Comment (Paulomart) [inline .github/workflows/ci.yml:171]: same as above

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -126,45 +129,196 @@ jobs:
...
+              type=gha,scope=${{github.ref_name}}-backend
+              type=gha,scope=${{github.base_ref}}-backend
+            cache-to: type=gha,mode=max,scope=${{github.ref_name}}-backend
+
+    e2e_ci:
+      name: Lint E2E
+      runs-on: ubuntu-latest
+      steps:
+        - name: Checkout
+          uses: actions/checkout@v3
```
Comment (Paulomart) [inline .github/workflows/ci.yml:182]: I think this job is named confusingly: There is no end to end testing here.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -126,45 +129,196 @@ jobs:
...
+            npm ci
+
+        - name: Lint
+          working-directory: e2e
+          run: |
+            npm run lint -- --max-warnings 0 --no-fix
+
+    e2e_test:
+      name: Run E2E Tests
+      needs: [e2e_ci,frontend_container,backend_container]
```
Comment (Paulomart) [inline .github/workflows/ci.yml:201]:
```suggestion
      needs: 
        - e2e_ci
        - frontend_container
        - backend_container
```

also why does this step depends on `e2e_ci`?

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -126,45 +129,196 @@ jobs:
...
+      runs-on: ubuntu-latest
+      services:
+        backend:
+          image: oeffis/backend:${{ github.sha }}-untested
+          ports:
+            - 3000:3000
+        frontend:
+          image: oeffis/frontend:${{ github.sha }}-untested
+          ports:
+            - 8000:80
```
Comment (Paulomart) [inline .github/workflows/ci.yml:211]: nice 👍

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -126,45 +129,196 @@ jobs:
...
+          uses: actions/upload-artifact@v3
+          with:
+            name: cypress-report
+            path: |
+              e2e/cypress/screenshots
+              e2e/cypress/videos
+
+    publish:
+      name: Publish Images
+      needs: [e2e_test]
```
Comment (Paulomart) [inline .github/workflows/ci.yml:242]:
```suggestion
      needs: 
        - e2e_test
```

imo lint should be _needed_ here

File: e2e/.gitignore [hunk 8]
```diff
@@ -0,0 +1 @@
+node_modules
```
Comment (Paulomart) [inline e2e/.gitignore:1]: this should be in the root .gitignore; no need to add it here again

**Review by Paulomart — COMMENTED** [review]
Awesome stuff, needs a few things otherwise good to go :)

File: .github/workflows/ci.yml (hunk 1 above)
Comment (timvahlbrock) [reply .github/workflows/ci.yml:90]: no, default is false. see https://github.com/docker/build-push-action#inputs

File: .github/workflows/ci.yml (hunk 5 above)
Comment (timvahlbrock) [reply .github/workflows/ci.yml:201]: to prohibit e2e tests that do not pass linting from running.

File: .github/workflows/ci.yml (hunk 7 above)
Comment (timvahlbrock) [reply .github/workflows/ci.yml:242]: you mean additionally? I disagree. I would leave the dependencies as they are. Publish doesn't need to know wether the files pass linting, but only wether they run succesfully.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (timvahlbrock) [reply .github/workflows/ci.yml:182]: suggestions?

**Review by Paulomart — APPROVED** [review]

File: e2e/.gitignore (hunk 8 above)
Comment (timvahlbrock) [reply e2e/.gitignore:1]: separate issue?

**Pull request merged**